controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -320,51 +324,66 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
-                // Method notifyTxnCreation ensures that notification completes
-                // even in the presence of n/w or segment store failures.
-                log.debug("Txn={}, notified segments stores", txnId));
-
-        // Step 5. Start tracking txn in timeout service
-        return notify.whenCompleteAsync((result, ex) -> {
-            int version = 0;
-            long executionExpiryTime = System.currentTimeMillis() + maxExecutionPeriod;
-            if (!txnFuture.isCompletedExceptionally()) {
-                version = txnFuture.join().getVersion();
-                executionExpiryTime = txnFuture.join().getMaxExecutionExpiryTime();
-            }
-            timeoutService.addTxn(scope, stream, txnId, version, lease, executionExpiryTime, scaleGracePeriod);
-            log.debug("Txn={}, added to timeout service on host={}", txnId, hostId);
-        }, executor).thenApplyAsync(v -> new ImmutablePair<>(txnFuture.join(), segmentsFuture.join()), executor);
+                    // Step 4. Notify segment stores about new txn.
+                    CompletableFuture<List<Segment>> segmentsFuture = txnFuture.thenComposeAsync(txnData ->
+                            streamMetadataStore.getActiveSegments(scope, stream, txnData.getEpoch(), ctx, executor), executor);
+
+                    CompletableFuture<Void> notify = segmentsFuture.thenComposeAsync(activeSegments ->
+                            notifyTxnCreation(scope, stream, activeSegments, txnId), executor).whenComplete((v, e) ->
+                            // Method notifyTxnCreation ensures that notification completes
+                            // even in the presence of n/w or segment store failures.
+                            log.debug("Txn={}, notified segments stores", txnId));
+
+                    // Step 5. Start tracking txn in timeout service
+                    return notify.whenCompleteAsync((result, ex) -> {
+                        int version = 0;
+                        long executionExpiryTime = System.currentTimeMillis() + maxExecutionPeriod;
+                        if (!txnFuture.isCompletedExceptionally()) {
+                            version = txnFuture.join().getVersion();
+                            executionExpiryTime = txnFuture.join().getMaxExecutionExpiryTime();
+                        }
+                        timeoutService.addTxn(scope, stream, txnId, version, lease, executionExpiryTime, scaleGracePeriod);
+                        log.debug("Txn={}, added to timeout service on host={}", txnId, hostId);
+                    }, executor).thenApplyAsync(v -> new ImmutablePair<>(txnFuture.join(), segmentsFuture.join()), executor);
+                }), e -> {
+            Throwable unwrap = Exceptions.unwrap(e);
+            return unwrap instanceof StoreException.WriteConflictException || unwrap instanceof StoreException.DataNotFoundException;
+        }, 5, executor));
```

Comment:
What is this `5` here? 

Can we try to break this method into multiple smaller ones. I had to search really hard to realize this was an argument to `retry`  some 50 lines above.

Reply:
refactored the method to extract methods from it.